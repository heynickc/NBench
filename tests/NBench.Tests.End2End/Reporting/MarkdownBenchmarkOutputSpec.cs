using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using NBench.Collection;
using NBench.Metrics;
using NBench.Metrics.Counters;
using NBench.Reporting;
using NBench.Reporting.Targets;
#if !CORECLR
using ApprovalTests;
using ApprovalTests.Reporters;
#endif
using NBench.Sdk;
using Xunit;
using Xunit.Abstractions;

namespace NBench.Tests.End2End.Reporting
{
#if !CORECLR
    [UseReporter(typeof(ApprovalTests.Reporters.DiffReporter))]
    public class MarkdownBenchmarkOutputSpec
    {
        private readonly ITestOutputHelper _output;
        private static readonly string _solutionPath =
            ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar
            + ".." + Path.DirectorySeparatorChar;
        private static readonly string _perfResultsPath =
            Path.GetFullPath(_solutionPath + "PerfResults");

        public MarkdownBenchmarkOutputSpec(ITestOutputHelper ouput)
        {
            _output = ouput;
        }

        [Fact]
        public void Should_compare_benchmark_output_static_data()
        {
            CleanPerfDir(_perfResultsPath);
            SystemTime.UtcNow = () => new DateTime(2017, 3, 13);

            var mdOutput = new MarkdownBenchmarkOutput(_perfResultsPath);
            var fakeBenchmarkResults = new BenchmarkFinalResults(
                new BenchmarkResults("NBench.FakeBenchmark",
                    new BenchmarkSettings(TestMode.Test, RunMode.Iterations, 30, 1000,
                        new List<IBenchmarkSetting>(),
                        new ConcurrentDictionary<MetricName, MetricsCollectorSelector>()),
                    new List<BenchmarkRunReport>()),
                new List<AssertionResult>());
            mdOutput.WriteBenchmark(fakeBenchmarkResults);
            var fakePerfResultsFile = Directory.GetFiles(_perfResultsPath, "NBench.FakeBenchmark*.md",
                SearchOption.AllDirectories);
            Approvals.Verify(File.ReadAllText(fakePerfResultsFile.FirstOrDefault()));
        }
        private static void CleanPerfDir(string path)
        {
            var files = Directory.GetFiles(path, "*.md", SearchOption.AllDirectories);
            foreach (var file in files)
            {
                File.Delete(file);
            }
        }
    }
#endif

}
