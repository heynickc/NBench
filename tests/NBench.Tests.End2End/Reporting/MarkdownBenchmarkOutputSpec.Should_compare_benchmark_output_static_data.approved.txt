# NBench.FakeBenchmark
_3/13/2017 12:00:00 AM_
### System Info
```ini
NBench=NBench, Version=0.3.4.0, Culture=neutral, PublicKeyToken=null
OS=Microsoft Windows NT 10.0.14393.0
ProcessorCount=8
CLR=4.0.30319.42000,IsMono=False,MaxGcGeneration=2
```

### NBench Settings
```ini
RunMode=Iterations, TestMode=Test
NumberOfIterations=30, MaximumRunTime=00:00:01
Concurrent=False
Tracing=False
```

## Data
-------------------

### Totals
|          Metric |           Units |             Max |         Average |             Min |          StdDev |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |

### Per-second Totals
|          Metric |       Units / s |         Max / s |     Average / s |         Min / s |      StdDev / s |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |

### Raw Data

